Merge upstream prometheus/prometheus at d0ecf9b

charleskorn opened the pull request:
This includes the following changes, none of which seem to be worthy of concern for Mimir:


Performance improvements:
* https://github.com/prometheus/prometheus/pull/13742: performance improvement to chunk checksum checks
* https://github.com/prometheus/prometheus/pull/13673: performance improvement to `Labels.String()`

Bug fixes:
* https://github.com/prometheus/prometheus/pull/13667: bug fixes for query statistics when native histograms are queried
* https://github.com/prometheus/prometheus/pull/13777
* https://github.com/prometheus/prometheus/pull/9633

Refactoring and tidy up:
* https://github.com/prometheus/prometheus/pull/13713: introduce `promql.QueryEngine` interface  and allow using it for rule evaluation
* https://github.com/prometheus/prometheus/pull/13763
* https://github.com/prometheus/prometheus/pull/13764

Changes that don't affect Mimir:
* https://github.com/prometheus/prometheus/pull/13772
* https://github.com/prometheus/prometheus/pull/12933
* https://github.com/prometheus/prometheus/pull/13099

Non-production changes (tests, comments, docs etc.):
* https://github.com/prometheus/prometheus/pull/13751: changes some tests to use `http.StatusXXX` constants instead of literals
* https://github.com/prometheus/prometheus/pull/13743
* https://github.com/prometheus/prometheus/pull/13728: upgrades prometheus/common in an example app
* https://github.com/prometheus/prometheus/pull/13752, https://github.com/prometheus/prometheus/pull/13770, https://github.com/prometheus/prometheus/pull/13768, https://github.com/prometheus/prometheus/pull/13783: add automation to publish container image readme to Docker Hub and Quay.io
* https://github.com/prometheus/prometheus/pull/13759: fix import path comments
* https://github.com/prometheus/prometheus/pull/13771

prymitive committed:
Fix BenchmarkScrapeLoopAppendOM

OpenMetrics requires EOF comment at the end of metrics body, but the makeTestMetrics() function doesn't append it.
This means this benchmark tests a response with errors but I don't think that was the intention.

Signed-off-by: Łukasz Mierzwa <l.mierzwa@gmail.com>

prymitive committed:
Check for duplicated series on a scrape

When Prometheus scrapes a target and it sees the same time series repeated multiple times it currently silently ignores that.
This change adds a test for that and fixes the scrape loop so that:

- Only first sample for each unique time series is appended
- Duplicated samples increment the prometheus_target_scrapes_sample_duplicate_timestamp_total metric

This allows one to identify such scrape jobs and targets.

Benchmark results:

```
name                            old time/op    new time/op    delta
ScrapeLoopAppend-8                64.8µs ± 2%    71.1µs ±20%   +9.75%  (p=0.000 n=10+10)
ScrapeLoopAppendOM-8              64.2µs ± 1%    68.5µs ± 7%   +6.71%  (p=0.000 n=9+10)
TargetsFromGroup/1_targets-8      14.2µs ± 1%    14.5µs ± 1%   +1.99%  (p=0.000 n=10+10)
TargetsFromGroup/10_targets-8      149µs ± 1%     152µs ± 1%   +2.05%  (p=0.000 n=9+10)
TargetsFromGroup/100_targets-8    1.49ms ± 4%    1.48ms ± 1%     ~     (p=0.796 n=10+10)

name                            old alloc/op   new alloc/op   delta
ScrapeLoopAppend-8                19.9kB ± 1%    17.8kB ± 3%  -10.23%  (p=0.000 n=8+10)
ScrapeLoopAppendOM-8              19.9kB ± 1%    18.3kB ±10%   -8.14%  (p=0.001 n=9+10)
TargetsFromGroup/1_targets-8      2.43kB ± 0%    2.43kB ± 0%   -0.15%  (p=0.045 n=10+10)
TargetsFromGroup/10_targets-8     24.3kB ± 0%    24.3kB ± 0%     ~     (p=0.083 n=10+9)
TargetsFromGroup/100_targets-8     243kB ± 0%     243kB ± 0%     ~     (p=0.720 n=9+10)

name                            old allocs/op  new allocs/op  delta
ScrapeLoopAppend-8                  9.00 ± 0%      9.00 ± 0%     ~     (all equal)
ScrapeLoopAppendOM-8                10.0 ± 0%      10.0 ± 0%     ~     (all equal)
TargetsFromGroup/1_targets-8        40.0 ± 0%      40.0 ± 0%     ~     (all equal)
TargetsFromGroup/10_targets-8        400 ± 0%       400 ± 0%     ~     (all equal)
TargetsFromGroup/100_targets-8     4.00k ± 0%     4.00k ± 0%     ~     (all equal)
```

Signed-off-by: Łukasz Mierzwa <l.mierzwa@gmail.com>

prymitive committed:
Fix TestScrapeLoopDiscardDuplicateLabels test

This test calls Rollback() which is normally called from within append code.
Doing so means that staleness tracking data is outdated and need to by cycled manually.

Signed-off-by: Łukasz Mierzwa <l.mierzwa@gmail.com>

prymitive committed:
Move staleness tracking out of checkAddError() calls

This call bloats checkAddError signature and logic, we can and should call it from the main scrape logic.

Signed-off-by: Łukasz Mierzwa <l.mierzwa@gmail.com>

beorn7 committed:
promql: remove redundant line

Signed-off-by: beorn7 <beorn@grafana.com>

prymitive committed:
Check of duplicated samples directly in scrapeCache.get()

Avoid extra map lookup by hooking check into cache get.

```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/scrape
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
                     │  main.txt   │              new.txt               │
                     │   sec/op    │   sec/op     vs base               │
ScrapeLoopAppend-8     66.72µ ± 0%   66.89µ ± 0%       ~ (p=0.879 n=50)
ScrapeLoopAppendOM-8   66.61µ ± 0%   66.89µ ± 1%       ~ (p=0.115 n=50)
geomean                66.66µ        66.89µ       +0.34%

                     │   main.txt    │               new.txt                │
                     │     B/op      │     B/op      vs base                │
ScrapeLoopAppend-8     20.17Ki ±  1%   20.12Ki ± 1%        ~ (p=0.343 n=50)
ScrapeLoopAppendOM-8   20.38Ki ± 10%   17.99Ki ± 2%  -11.69% (p=0.017 n=50)
geomean                20.27Ki         19.03Ki        -6.14%

                     │  main.txt  │               new.txt               │
                     │ allocs/op  │ allocs/op   vs base                 │
ScrapeLoopAppend-8     11.00 ± 0%   11.00 ± 0%       ~ (p=1.000 n=50) ¹
ScrapeLoopAppendOM-8   12.00 ± 0%   12.00 ± 0%       ~ (p=1.000 n=50) ¹
geomean                11.49        11.49       +0.00%
¹ all samples are equal
```

Signed-off-by: Łukasz Mierzwa <l.mierzwa@gmail.com>

beorn7 committed:
promql: Add code comment

Signed-off-by: beorn7 <beorn@grafana.com>

beorn7 committed:
promql: Add histograms to TestQueryStatistics

Also, fix the bugs exposed by the tests.

Signed-off-by: beorn7 <beorn@grafana.com>

charleskorn committed:
Allow using alternative PromQL engines for rule evaluation

Signed-off-by: Charles Korn <charles.korn@grafana.com>

carrychair committed:
fix function and struct name

Signed-off-by: carrychair <linghuchong404@gmail.com>

charleskorn committed:
Remove unnecessary `SetQueryLogger` method on `QueryEngine` interface

Signed-off-by: Charles Korn <charles.korn@grafana.com>

testwill committed:
chore: use constant instead of numeric literal

Signed-off-by: guoguangwu <guoguangwug@gmail.com>

Commit:
Merge pull request #13751 from testwill/http_statuscode

chore: use constant instead of numeric literal

Commit:
Merge pull request #13743 from carrychair/main

fix function and struct names in comments

bboreham committed:
[ENHANCEMENT] TSDB: Check CRC without allocating (#13742)

Use the existing utility function which does this.

Signed-off-by: Bryan Boreham <bjboreham@gmail.com>

bboreham committed:
labels: optimize String method (#13673)

Use a stack buffer to reduce memory allocations.

`Write(AppendQuote(AvailableBuffer` does not allocate or copy when
the buffer has sufficient space.

Also add a benchmark, with some refactoring.

Signed-off-by: Bryan Boreham <bjboreham@gmail.com>

SuperQ committed:
Add GitHub action to publish container README

Add a GitHub action to publish the README.md to Docker Hub and Quay.io.

Fixes: https://github.com/prometheus/prometheus/issues/5348

Signed-off-by: SuperQ <superq@gmail.com>

jcajka committed:
otlptranslator: fix up import paths

Signed-off-by: Jakub Čajka <jcajka@redhat.com>

beorn7 committed:
Merge pull request #13667 from prometheus/beorn7/promql

Improve TestQueryStatistics and fix bugs exposed by it

dependabot[bot] committed:
build(deps): bump github.com/prometheus/common (#13728)

Bumps [github.com/prometheus/common](https://github.com/prometheus/common) from 0.49.0 to 0.50.0.
- [Release notes](https://github.com/prometheus/common/releases)
- [Commits](https://github.com/prometheus/common/compare/v0.49.0...v0.50.0)

---
updated-dependencies:
- dependency-name: github.com/prometheus/common
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

beorn7 committed:
Merge pull request #13752 from prometheus/superq/publish_docker_readme

Add GitHub action to publish container README

SuperQ committed:
Fix container_description workflow

Fix yaml indentation. :facepalm:

Signed-off-by: SuperQ <superq@gmail.com>

bwplotka committed:
Merge pull request #13713 from charleskorn/query-engine-interface

rules: allow using alternative PromQL engines for rule evaluation by callers using Prometheus as a lib.

Commit:
Chunked remote read: close the querier earlier

I have seen prometheis instances misebehaving because of broken chinked remote
read requests.

In order to avoid OOM's when this happens, I propose to close the
queries used by the streamed remote read requests earlier.

Signed-off-by: Julien Pivotto <roidelapluie@o11y.eu>

SuperQ committed:
Normalize and fixup step names.

Signed-off-by: SuperQ <superq@gmail.com>

bboreham committed:
[Cleanup] TSDB: Remove old deprecated WAL implementation

Deprecated since 2018.

Signed-off-by: Bryan Boreham <bjboreham@gmail.com>

charleskorn committed:
Only run on prometheus/prometheus repo

Signed-off-by: Charles Korn <charles.korn@grafana.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/mimir-prometheus?pullRequest=607) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/mimir-prometheus?pullRequest=607) before we can accept your contribution.<br/>**7** out of **15** committers have signed the CLA.<br/><br/>:white_check_mark: beorn7<br/>:white_check_mark: testwill<br/>:white_check_mark: bboreham<br/>:white_check_mark: gouthamve<br/>:white_check_mark: jkroepke<br/>:white_check_mark: charleskorn<br/>:white_check_mark: aknuds1<br/>:x: prymitive<br/>:x: carrychair<br/>:x: bwplotka<br/>:x: jcajka<br/>:x: Julien<br/>:x: machine424<br/>:x: darshanime<br/>:x: SuperQ<br/><hr/>**Julien** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/mimir-prometheus?pullRequest=607) it.</sub>

aknuds1 commented on .github/workflows/container_description.yml:
```diff
@@ -0,0 +1,55 @@
+---
+name: Push README to Docker Hub
```
Do we want this workflow in the fork? This would be the only workflow shared with upstream. AFAICT, you make this workflow essentially no-op, by checking that the repository is prometheus/prometheus. Why not follow the pattern of the other workflows for consistency, and leave them out?

charleskorn replied:
We don't want this workflow in any fork - https://github.com/prometheus/prometheus/pull/13768 will disable it for forks.

To me, it seems easier to take the workflow file as-is including the condition so that it won't run in our fork, rather than removing the workflow altogether. Then we don't have to worry about conflicts etc. in the future.

If you're OK with this approach, I'll wait until that PR is merged and then update this PR to include all changes up to and including it.

aknuds1 replied:
Thanks for the explanation, sounds good to me @charleskorn.

SuperQ committed:
Merge pull request #13761 from prometheus/superq/fixup_workflow

Fix container_description workflow

aknuds1 reviewed (commented):
Generally LGTM, just wondering about why this new GitHub workflow is included (while we don't share any other workflows with upstream).

SuperQ committed:
Add container_description.yml to repo sync

Add the container_description.yml workflow to the repo file sync script.
* Skip sync if there is no Dockerfile.
* Fixup minor typo in container_description.yml.

Signed-off-by: SuperQ <superq@gmail.com>

aknuds1 committed:
Remove unused function tsdb/chunks.PopulatedChunk (#13763)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

SuperQ committed:
Merge pull request #13770 from prometheus/superq/sync_description

Add container_description.yml to repo sync

machine424 committed:
adjust signal termination warning log

Signed-off-by: machine424 <ayoubmrini424@gmail.com>

beorn7 committed:
Merge pull request #13772 from machine424/signals

adjust signal termination warning log

gouthamve committed:
Merge pull request #13759 from jcajka/main

otlptranslator: fix up import paths

bboreham committed:
TSDB: remove unused function

Was only used in old WAL implementation.

Signed-off-by: Bryan Boreham <bjboreham@gmail.com>

bboreham committed:
TSDB: move function only used in tests

Signed-off-by: Bryan Boreham <bjboreham@gmail.com>

Commit:
Merge pull request #13777 from roidelapluie/remoteread2

Chunked remote read: close the querier earlier

bboreham committed:
MAINTAINERS: Add Bryan Boreham, Ayoub Mrini (#13771)

Also simplify structure. Ordering of 'general' maintainers is alphabetical by 2nd name.

Signed-off-by: Bryan Boreham <bjboreham@gmail.com>

darshanime committed:
Fix retention boundary so 2h retention deletes blocks right at the 2h boundary (#9633)

Signed-off-by: darshanime <deathbullet@gmail.com>

bboreham committed:
Merge pull request #12933 from prymitive/duplicated_samples

When Prometheus scrapes a target and it sees the same time series repeated multiple times it currently silently ignores that. This change adds a test for that and fixes the scrape loop so that:

* Only first sample for each unique time series is appended
* Duplicated samples increment the prometheus_target_scrapes_sample_duplicate_timestamp_total metric
This allows one to identify such scrape jobs and targets.

Also fix some tests and benchmark.

jkroepke committed:
{discovery,remote_write}/azure: Support default SDK authentication (#13099)

* discovery/azure: Offer default SDK authentication

Signed-off-by: Jan-Otto Kröpke <mail@jkroepke.de>

bboreham committed:
Merge pull request #13764 from bboreham/remove-deprecated-wal

[Cleanup] TSDB: Remove old deprecated WAL implementation

Deprecated since 2018.

charleskorn committed:
Support copying workflow as-is to other Prometheus repositories.

Signed-off-by: Charles Korn <charles.korn@grafana.com>

SuperQ committed:
Merge pull request #13768 from charleskorn/readme-publish-workflow

Signed-off-by: Ben Kochie <superq@gmail.com>

SuperQ committed:
Add container_description.yml to force sync list

Since we skip repos that don't have a Dockerfile, we can force sync the
`.github/workflows/container_description.yml` config.

Signed-off-by: SuperQ <superq@gmail.com>

Commit:
Merge pull request #13783 from prometheus/superq/sync_fixup

Add container_description.yml to force sync list

charleskorn committed:
Merge branch 'main' into charleskorn/merge-upstream

aknuds1 reviewed (approved):
LGTM, thanks!
